draw hooks

ldstein opened the pull request:
Hi @leeoniya,

This PR proposes adding `willPaint` and `didPaint` hooks to `Line`.

It will enable custom rendering onto the Canvas element, either before or after the Chart is drawn.

I added this functionality as an approach to add highlighted annotated regions of a Chart (BG color below, Text labels above). 

If you feel it's a worthwhile addition, I'll add some documentation to `readme.md` accordingly.

**Edit** I discovered I could achieve a similar result using overlayed/underlayed canvas / div elements and rely on `setScale`,  `setFocus` for notification of change. I'll leave the PR open regardless for your consideration.

![image](https://user-images.githubusercontent.com/3657344/71347101-7e2d8c80-25a4-11ea-9a07-1af543205a2d.png)

leeoniya commented:
thanks for the PR, this is cool :)

there's actually a couple discussions to be had here.

1. generic paint/draw hooks
2. range annotations & this demo

**1. generic paint/draw hooks**

generally, i think having generic draw hooks is a good idea &  escape hatch. while i'm no stranger to will/did hook implementations [1], i think having this style just for draw hooks is not ideal. effectively, all of uPlot's current hooks are `did*` hooks. additionally, i think it would be better to have more than just two but to make it possible to draw into any series' "layer", not just above or below all of them. so maybe something like:

- `drawClear` fires after the canvas is cleared
- `drawGrid` fires after the grid is drawn
- `drawSeries: (self, i)` (fires after each series is drawn)
- `draw` fires after everything is drawn

rather than passing `ctx` into the hooks i'd rather follow the existing convention and simply expose `ctx` as `self.ctx` (since it's a singleton). hooks that accept extra params are the ones that need further context (like "which" in setSeries and setScale).

for this demo specifically, you should not need to replicate getXPos & getYPos - they're already exposed:

https://github.com/leeoniya/uPlot/blob/589d1dd2a17d7fdc33aef176b837678131b460c2/dist/uPlot.cjs.js#L1672-L1674

**2. range annotations & this demo**

i think that chart annotations provide sufficient data inference utility to be part of the core, so there's https://github.com/leeoniya/uPlot/issues/27 already open to discuss the possible API. i'll comment a bit further there.

[1] https://github.com/domvm/domvm#lifecycle-hooks

ldstein commented:
Thanks for the feedback, all valid points.

I updated to reflect your input:
- Removed `willPaint`, `didPaint`
- Added `drawClear`, `drawGrid`, `drawSeries`, `draw` hooks
- Updated demo html

`draw-hooks.html` demonstrates various effects which can be applied, such as adding blur to the grid:

![image](https://user-images.githubusercontent.com/3657344/71392942-0c594f80-2645-11ea-94eb-fea597f1cb15.png)

Regarding **"you should not need to replicate getXPos & getYPos"**, there is a small offset issue when using `valToPos()` or `posToVal()` if the user's browser is set to a zoom level other than 100%.

Appears related to `valToPos()` using `cssWidth` instead of `canvas.width` when calculating position. Calculated correctly for DOM elements, but missing an offset when applied to canvas space. 

Hence why I copied `getXPos` & `getYPos` in the original demo :) Here is an example of the issue:

![image](https://user-images.githubusercontent.com/3657344/71393764-12e9c600-2649-11ea-83fb-53bce320162c.png)

**Other Draw Event Candidates**
- `drawGridAxis` might be another valid hook worth considering.
- drawPath() could potentially fire `drawSeriesStroke`, `drawSeriesFill`

leeoniya commented on src/Line.js:
```diff
@@ -717,7 +722,8 @@ export function Line(opts, data) {
 			}
-		});
+        });
+        fire("drawGrid");
```
let's move this up into `if (grid.show) {` above. it'd be weird if `drawGrid` fired when there's no grid to draw.

ldstein replied:
Maybe `fire("drawAxis", axis)` would be more appropriate in that case? 

Can access grid from `axis.grid`

leeoniya replied:
yep, good catch.

leeoniya replied:
err, sorry. i meant `fire("drawGrid", i)`. the axes are not really "drawn" on the canvas (they're in the dom). so let's stick with canvas terminology.

ldstein replied:
I changed this to `fire('drawGridAxis', a)` for each axis. If 1 or more axis are drawn, `drawGrid` will be fired. 

Allows hooking in directly after all axis have been drawn and before series is drawn.

leeoniya commented on src/Line.js:
```diff
@@ -536,8 +538,11 @@ export function Line(opts, data) {
-				drawPath(i);
+			if (i > 0 && s.show) {
+                drawPath(i);
+                fire("drawSeries", s, i);
```
this should just be `fire("drawSeries", i);` the `s` can already be pulled from `self.series.y[i+1]`

leeoniya commented on src/Line.js:
```diff
@@ -742,10 +748,12 @@ export function Line(opts, data) {
+        fire("drawClear");
+        drawAxesGrid();        
+        drawSeries();
+        fire("draw");
```
i'd probably move this down 1 line (below `didPaint = true;`)

leeoniya commented:
> Regarding "you should not need to replicate getXPos & getYPos", there is a small offset issue when using valToPos() or posToVal() if the user's browser is set to a zoom level other than 100%.
>
> Appears related to valToPos() using cssWidth to calculate position. It calculates correctly for DOM space, but missing a scaling modifier when applied to canvas space.
>
> Hence why I copied getXPos & getYPos in the original demo :) Here is an example of the issue:

you need to multiply the position by `devicePixelRatio`. this solves the issue on high res displays as well as higher zoom levels:

```js
let yPos0 = u.valToPos(maxVal, 'y') * devicePixelRatio;
let yPos1 = u.valToPos(minVal, 'y') * devicePixelRatio;
```

> Other Draw Event Candidates

honestly, i don't feel the value of those is that great.

please switch to tabs spacing and clean up all the superfluous whitespace changes, then maybe squash the commits. otherwise this looks good!

leeoniya commented on demos/draw-hooks.html:
```diff
@@ -0,0 +1,231 @@
+                    ],
+                    drawSeries: [
+                        enableBlur,
+                        drawSeriesBoundary,
```
i didn't actually realize what this was doing until i disabled it. until that point it looked like it was just part of the a blurred gradient background.

ldstein replied:
Yeah, not a particularly great example. It's more apparent once you zoom in (then you get a bounding box around the visible points).

leeoniya commented on demos/draw-hooks.html:
```diff
@@ -0,0 +1,237 @@
+                    ctx.closePath();                    
+                }
+
+                function setCtxProperties(ctx, settings)
```
i'm not a canvas expert, but isn't this effectively https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/save?

leeoniya commented on src/Line.js:
```diff
@@ -540,7 +540,7 @@ export function Line(opts, data) {
 			if (i > 0 && s.show) {
                 drawPath(i);
-                fire("drawSeries", s, i);
+                fire("drawSeries", i-1);
```
right now i'm passing the raw internal `i` here [1], so this should not be adjusted to assume differently. whether this is ultimately a good decision is probably another topic, but let's keep it uniform in this PR.

[1] https://github.com/leeoniya/uPlot/blob/master/src/Line.js#L929

leeoniya replied:
i've opened #77 to figure this out.

leeoniya commented:
thanks for contributing! i did some cleanup commits.

ldstein commented:
Thanks @Leeyon,

I meant to comment earlier, but you Christmas kind of took priority :)

Having experimented a bit now, the real-life requirements I have can be accomplished with just the `drawClear` and `draw` hooks. 

It may be enough for 80% of use cases and maintains the "black box" around Uplot's internal drawing routines:

--------------------------------------------
"drawClear" Any custom drawing here
--------------------------------------------
uPlot drawing methods (controlled via options) 
--------------------------------------------
"draw" Any custom drawing here
--------------------------------------------

As you already mentioned,  #27 potentially competes with 'drawSeries'.
Leaving out 'drawSeries' also means no need to solve #77 right away.

Also note there is a bug in the demo I provided:

When `drawClear` fires, I enable a canvas blur filter (in preparation of drawing the grid). I then turn this off by when `drawGrid` is fired and checking if the axis being drawn is `y`.

However, `drawGrid` for the y axis does not fire all the time (for example, when zooming in a great distance). The result is the blur filter is never be turned off, so the entire chart is blurred.

A `drawGrids/drawAllGrids` (which is always fired regardless of how many grids were drawn) would allow hooking into directly before the series are drawn and would solve this particularly issue. 

Again, I don't think I'd miss it if the `drawGrid` hook didn't exist. I could always configure the grid with semi-transparent colors and use `drawClear` to do something fancy.

Food for thought.

ldstein commented:
Thanks @Leeyon,

I meant to comment earlier, but you Christmas kind of took priority :)

Having experimented a bit now, the real-life requirements I have can be accomplished with just the `drawClear` and `draw` hooks. 

It may be enough for 80% of use cases and maintains the "black box" around Uplot's internal drawing routines:

--------------------------------------------
"drawClear" Any custom drawing here
--------------------------------------------
uPlot drawing methods (controlled via options) 
--------------------------------------------
"draw" Any custom drawing here
--------------------------------------------

As you already mentioned,  #27 potentially competes with 'drawSeries'.
Leaving out 'drawSeries' also means no need to solve #77 right away.

Also note there is a bug in the demo I provided:

When `drawClear` fires, I enable a canvas blur filter (in preparation of drawing the grid). I then turn this off by when `drawGrid` is fired and checking if the axis being drawn is `y`.

However, `drawGrid` for the y axis does not fire all the time (for example, when zooming in a great distance). The result is the blur filter is never be turned off, so the entire chart is blurred.

A `drawGrids/drawAllGrids` (which is always fired regardless of how many grids were drawn) would allow hooking into directly before the series are drawn and would solve this particularly issue. 

Again, I don't think I'd miss it if the `drawGrid` hook didn't exist. I could always configure the grid with semi-transparent colors and use `drawClear` to do something fancy.

Food for thought.

leeoniya commented:
> I meant to comment earlier, but Christmas kind of took priority :)

of course, merry christmas, btw!

> As you already mentioned, #27 potentially competes with `drawSeries`.

i think this `draw-hooks` demo (not `drawSeries`) "competes" with #27, (and also #10). but while those features are absent, it's a good thing to show; once they're added, the this demo will need to be changed to something that doesn't overlap those APIs.

> Having experimented a bit now, the real-life requirements I have can be accomplished with just the `drawClear` and `draw` hooks.

but it's pretty cheap to have the others, too. and without them it's impossible to draw into each series' layer and allow those to be visually occluded by later series.

> Omitting `drawSeries` also means no need to solve #77 right away.

#77 also affects `setSeries` hooks, so it's a priority for me before 1.0. the current plan is to get rid of the x/y layers and just have flat arrays for `series` and `axes`. for series, it makes perfect sense because they always need 1:1 mapping to the data array. for axes, `side` is really the thing that determines where they're drawn, so `x` and `y` nesting them is just unnecessary indirection.

when both become arrays, the rules are simple. `axes[0]` defaults to 'x' and `axes[1..N]` default to `y` options. same goes for `series`. this means the opts will always be:

```js
{
  series: [
    {},
    {color: "red"},
  ],
  axes: [
    {},
    {values: ...},
  ]
}
```

sadly, it's not a beautiful API, but it resolves the `i` situation without any mental gymnastics or additional conventions. that's kind of where i'm at with that.

> Also note there is a bug in the demo I provided:

feel free to submit any improvements, but the demos don't need to 100% bug free, more just starting points for more experimentation. i'd rather have imperfect but simple demos rather than demos which handle all edge cases but are more difficult to grok. i think leaving a comment about the bug is fine, especially if the alternative is another 20 lines of code or more state-tracking complexity.

ldstein commented:
No worries, sounds good. Thanks again for all your effort, and Merry Christmas to you too!

Fiv5 commented:
Nice conversation, I learn a lot!
